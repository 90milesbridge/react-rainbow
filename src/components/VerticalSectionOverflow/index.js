import React, { Component } from 'react';
import PropTypes from 'prop-types';
import classnames from 'classnames';
import IconSvg from './../IconSvg';
import Label from './label';
import AssistiveText from './../AssistiveText';
import { uniqueId } from './../../libs/utils';

const searchResultsId = uniqueId('search-results');

/**
* Represents an overflow of items from a preceding verticalNavigationSection,
* with the ability to toggle visibility.
*/
export default class VerticalSectionOverflow extends Component {
    constructor(props) {
        super(props);
        this.state = {
            isExpanded: props.expanded,
        };
        this.toggleOverflow = this.toggleOverflow.bind(this);
    }

    getContainerClassNames() {
        const { className } = this.props;
        return classnames('slds-nav-vertical__overflow', className);
    }

    getOverflowClassName() {
        const { isExpanded } = this.state;
        if (isExpanded) {
            return 'slds-show';
        }
        return 'slds-hide';
    }

    toggleOverflow() {
        const { isExpanded } = this.state;
        this.setState({ isExpanded: !isExpanded });
    }

    render() {
        const {
            expandedLabel,
            collapsedLabel,
            style,
            assistiveText,
            children,
        } = this.props;
        const { isExpanded } = this.state;

        return (
            <div className={this.getContainerClassNames()} style={style}>
                <button
                    className="slds-button slds-button_reset slds-nav-vertical__action slds-nav-vertical__action_overflow"
                    aria-controls={searchResultsId}
                    aria-expanded={isExpanded}
                    onClick={this.toggleOverflow}>

                    <IconSvg iconName="utility:chevronright" className="slds-button__icon slds-button__icon_left" />
                    <span className="slds-nav-vertical__action-text">
                        <Label
                            isExpanded={isExpanded}
                            expandedLabel={expandedLabel}
                            collapsedLabel={collapsedLabel} />

                        <AssistiveText text={assistiveText} />
                    </span>

                </button>
                <div data-id="vertical-overflow" id={searchResultsId} className={this.getOverflowClassName()}>
                    <ul>
                        {children}
                    </ul>
                </div>
            </div>
        );
    }
}

VerticalSectionOverflow.propTypes = {
    /** The label to show when the section is collapsed. */
    collapsedLabel: PropTypes.node,
    /** The label to show when the section is expanded. */
    expandedLabel: PropTypes.node,
    /** The state of the overflow. */
    expanded: PropTypes.bool,
    /** A description for assistive sreen readers. */
    assistiveText: PropTypes.string,
    /** A CSS class for the outer element, in addition to the component's base classes. */
    className: PropTypes.string,
    /** An object with custom style applied for the outer element. */
    style: PropTypes.object,
    /**
    * This prop that should not be visible in the documentation.
    * @ignore
    */
    children: PropTypes.node,
};

VerticalSectionOverflow.defaultProps = {
    expandedLabel: 'Show Less',
    collapsedLabel: 'Show More',
    expanded: false,
    className: undefined,
    style: undefined,
    assistiveText: undefined,
    children: null,
};
