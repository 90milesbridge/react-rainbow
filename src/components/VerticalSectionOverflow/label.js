import PropTypes from 'prop-types';

export default function Label({ expandedLabel, collapsedLabel, isExpanded }) {
    if (isExpanded) {
        return expandedLabel;
    }
    return collapsedLabel;
}

Label.propTypes = {
    expandedLabel: PropTypes.node.isRequired,
    collapsedLabel: PropTypes.node.isRequired,
    isExpanded: PropTypes.bool.isRequired,
};
