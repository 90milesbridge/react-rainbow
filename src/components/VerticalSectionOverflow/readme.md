##### vertical section overflow

    class ShadedVerticalNavigation extends React.Component {
        constructor(props) {
            super(props);
            this.state = {
                selectedItem: 'item 1',
            };
            this.handleOnSelect = this.handleOnSelect.bind(this);
        }

        handleOnSelect(e, selectedItem) {
            return this.setState({ selectedItem });
        }

        render() {
            return (
                <VerticalNavigation selectedItem={this.state.selectedItem} onSelect={this.handleOnSelect}>
                    <VerticalSection label="REPORTS">
                        <VerticalItem name="item-1" label="Recent" />
                        <VerticalItem name="item-2" label="Created by Me" />
                    </VerticalSection>
                    <VerticalSectionOverflow expandedLabel= "FOLDERS" collapsedLabel= "FOLDERS">
                        <VerticalItem name="item-3" label="Created by Me" />
                        <VerticalItem name="item-4" label="Shared with Me" />
                    </VerticalSectionOverflow>
                    <VerticalSectionOverflow expanded expandedLabel= "DOCUMENTS" collapsedLabel= "DOCUMENTS">
                        <VerticalItem name="item-5" label="Created by Me" />
                        <VerticalItem name="item-6" label="Shared with Me" />
                    </VerticalSectionOverflow>
                    <VerticalSection label="COMPONENTS">
                        <VerticalItem name="item-7" label="VerticalSectionOverflow" href="/#/Components/VerticalSectionOverflow" />
                        <VerticalItem name="item-8" label="VerticalNavigation" href="/#/Components/VerticalNavigation" />
                    </VerticalSection>
                </VerticalNavigation>
            );
        }
    }

    <div>
        <GlobalHeader />
        <div className="slds-large-size_1-of-3 slds-medium-size_1-of-2 slds-small-size_1-of-1 slds-color__background_gray-1 slds-p-vertical_medium slds-border_right">
            <ShadedVerticalNavigation />
        </div>
    </div>