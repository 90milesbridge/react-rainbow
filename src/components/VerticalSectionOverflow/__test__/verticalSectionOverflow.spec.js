import React from 'react';
import { mount } from 'enzyme';
import renderer from 'react-test-renderer';
import VerticalSectionOverflow from './../';

describe('<VerticalSectionOverflow/>', () => {
    it('should change the isExpanded state when the button element is clicked', () => {
        const component = mount(
            <VerticalSectionOverflow />,
        );
        expect(component.state().isExpanded).toBe(false);
        component.find('button').simulate('click');
        expect(component.state().isExpanded).toBe(true);
    });
    it('should change the isExpanded state when the button element is clicked and the expanded is set to true', () => {
        const component = mount(
            <VerticalSectionOverflow expanded />,
        );
        expect(component.state().isExpanded).toBe(true);
        component.find('button').simulate('click');
        expect(component.state().isExpanded).toBe(false);
    });
    it('should set a generated id as aria-controls in button element and as id in the overflow container', () => {
        const component = mount(
            <VerticalSectionOverflow />,
        );
        expect(component.find('button').prop('aria-controls')).toMatch(/search-results/);
        expect(component.find('div[data-id="vertical-overflow"]').prop('id')).toMatch(/search-results/);
    });
    it('should set the aria-expanded in the button element to false initially', () => {
        const component = mount(
            <VerticalSectionOverflow />,
        );
        expect(component.find('button').prop('aria-expanded')).toBe(false);
    });
    it('should set the aria-expanded in the button element to true if expanded is passed', () => {
        const component = mount(
            <VerticalSectionOverflow expanded />,
        );
        expect(component.find('button').prop('aria-expanded')).toBe(true);
    });
    it('should set the aria-expanded in the button element to true if it is clicked', () => {
        const component = mount(
            <VerticalSectionOverflow />,
        );
        expect(component.find('button').prop('aria-expanded')).toBe(false);
        component.find('button').simulate('click');
        expect(component.find('button').prop('aria-expanded')).toBe(true);
    });
    it('should pass the right props to the IconSvg component', () => {
        const component = mount(
            <VerticalSectionOverflow />,
        );
        expect(component.find('IconSvg').props()).toEqual({
            iconName: 'utility:chevronright',
            className: 'slds-button__icon slds-button__icon_left',
        });
    });
    it('should pass the right props to Label component when use default values', () => {
        const component = mount(
            <VerticalSectionOverflow />,
        );
        expect(component.find('Label').props()).toEqual({
            collapsedLabel: 'Show More',
            expandedLabel: 'Show Less',
            isExpanded: false,
        });
    });
    it('should pass the right props to Label component when the respective props are passed', () => {
        const component = mount(
            <VerticalSectionOverflow collapsedLabel="collapsed label" expandedLabel="expanded label" expanded />,
        );
        expect(component.find('Label').props()).toEqual({
            collapsedLabel: 'collapsed label',
            expandedLabel: 'expanded label',
            isExpanded: true,
        });
    });
    it('should set the isExpanded prop in Label component to true when the the button is clicked', () => {
        const component = mount(
            <VerticalSectionOverflow />,
        );
        expect(component.find('Label').prop('isExpanded')).toBe(false);
        component.find('button').simulate('click');
        expect(component.find('Label').prop('isExpanded')).toBe(true);
    });
    it('should set the assistiveText passed as text in AssistiveText component', () => {
        const component = mount(
            <VerticalSectionOverflow assistiveText="description text" />,
        );
        expect(component.find('AssistiveText').prop('text')).toBe('description text');
    });
    it('should set the right className in the overflow container when is closed', () => {
        const component = mount(
            <VerticalSectionOverflow />,
        );
        expect(component.find('div[data-id="vertical-overflow"]').prop('className')).toBe('slds-hide');
    });
    it('should set the right className in the overflow container when expanded is passed', () => {
        const component = mount(
            <VerticalSectionOverflow expanded />,
        );
        expect(component.find('div[data-id="vertical-overflow"]').prop('className')).toBe('slds-show');
    });
    it('should set the right className in the overflow container when the button is clicked', () => {
        const component = mount(
            <VerticalSectionOverflow />,
        );
        expect(component.find('div[data-id="vertical-overflow"]').prop('className')).toBe('slds-hide');
        component.find('button').simulate('click');
        expect(component.find('div[data-id="vertical-overflow"]').prop('className')).toBe('slds-show');
    });
    it('should render the children passed', () => {
        const component = mount(
            <VerticalSectionOverflow>
                <span>the section children</span>
            </VerticalSectionOverflow>,
        );
        expect(component.find('ul').text()).toBe('the section children');
    });
    it('should have the right class names when a custom className is passed', () => {
        const component = renderer.create(
            <VerticalSectionOverflow className="custom-class-name" />,
        );
        expect(component).toMatchSnapshot();
    });
});
