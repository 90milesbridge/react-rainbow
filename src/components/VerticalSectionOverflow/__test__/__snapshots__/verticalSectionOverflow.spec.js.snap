// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<VerticalSectionOverflow/> should have the right class names when a custom className is passed 1`] = `
<div
  className="slds-nav-vertical__overflow custom-class-name"
  style={Object {}}
>
  <button
    aria-controls="search-results-1"
    aria-expanded={false}
    className="slds-button slds-button_reset slds-nav-vertical__action slds-nav-vertical__action_overflow"
    onClick={[Function]}
  >
    <svg
      aria-hidden={true}
      className="slds-button__icon slds-button__icon_left"
      style={Object {}}
    >
      <use
        xlinkHref="/icons/utility-sprite/svg/symbols.svg#chevronright"
        xmlnsXlink="http://www.w3.org/1999/xlink"
      />
    </svg>
    <span
      className="slds-nav-vertical__action-text"
    >
      Show More
    </span>
  </button>
  <div
    className="slds-hide"
    data-id="vertical-overflow"
    id="search-results-1"
  >
    <ul />
  </div>
</div>
`;
